cmd/entrypoint/webhooks.go

```diff
@@ -0,0 +1,312 @@
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	k8sTypesMetaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	k8sClientCoreV1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/conversion"
+
+	"github.com/datawire/ambassador/v2/pkg/api/getambassador.io/v2"
+	"github.com/datawire/ambassador/v2/pkg/api/getambassador.io/v3alpha1"
+	"github.com/datawire/ambassador/v2/pkg/k8s"
+	"github.com/datawire/dlib/derror"
+	"github.com/datawire/dlib/dhttp"
+)
+
+const (
+	webhookPath   = "/crdconvert"
+	certValidDays = 365
+	caSecretName  = "emissary-ingress-webhook-ca"
+)
+
+// TODO: automatic cert regeneration
+
+func constStringPtr(x string) *string {
+	return &x
+}
+
+func int32Ptr(x int) *int32 {
+	y := int32(x)
+	return &y
+}
+
+func GetEmissaryScheme() *runtime.Scheme {
+	scheme := runtime.NewScheme()
+
+	utilruntime.Must(v2.AddToScheme(scheme))
+	utilruntime.Must(v3alpha1.AddToScheme(scheme))
+	return scheme
+}
+
+func HandleWebhooks(ctx context.Context, webhookPort int, scheme *runtime.Scheme) error {
+	// Create the webhook server
+	webhook := &conversion.Webhook{}
+	if err := webhook.InjectScheme(scheme); err != nil {
+		return err
+	}
+	mux := http.NewServeMux()
+	mux.HandleFunc(webhookPath, webhook.ServeHTTP)
+
+	// need a k8s client
+	kubeinfo := k8s.NewKubeInfo("", "", "")
+	restConfig, err := kubeinfo.GetRestConfig()
+	if err != nil {
+		return err
+	}
+	coreClient, err := k8sClientCoreV1.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	// get CA secret
+	secrets := coreClient.Secrets(GetAmbassadorNamespace())
+	caSecret, err := secrets.Get(ctx, caSecretName, k8sTypesMetaV1.GetOptions{})
+	var caPEM *b.Buffer
+	var caTemplate *x509.Certificate
+	var caPrivKey *rsa.PrivateKey
+	if err != nil {
+		// Error here most likely means the secret doesnt exist yet
+		// (then we make one). Otherwise if its an actual error we exit here
+		if !k8sErrors.IsNotFound(err) {
+			return err
+		}
+
+		// Generate CA Certificate and key...
+		notBefore := time.Now()
+		notAfter := notBefore.Add(time.Duration(certValidDays*24) * time.Hour)
+		serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+		serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+		if err != nil {
+			return err
+		}
+		caTemplate = &x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				Organization: []string{"Ambassador Labs"},
+			},
+			NotBefore:             notBefore,
+			NotAfter:              notAfter,
+			IsCA:                  true,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			BasicConstraintsValid: true,
+		}
+		caPrivKey, err = rsa.GenerateKey(rand.Reader, 4096)
+		if err != nil {
+			return err
+		}
```

Comment:
This is long enough that it probably makes sense to factor out separate `genKey` and `genCert` functions or something.